### PR TITLE
free rapids buffers safely if an exception is thrown

### LINKED ISSUE #2348: [BUG] Close called too many times DeviceMemoryBuffer in JoinGatherer
**Describe the bug**
When running against branch-0.6 (commit fdf585e) on TPC-DS queries, getting this exception.

**Steps/Code to reproduce bug**
Spark shell command:
```
/opt/spark/bin/spark-shell\
 --master spark://127.0.0.1:7077\
 --conf spark.serializer=org.apache.spark.serializer.KryoSerializer\
 --conf spark.kryoserializer.buffer=128m\
 --conf spark.kryo.registrator=com.nvidia.spark.rapids.GpuKryoRegistrator\
 --conf spark.locality.wait=0s\
 --conf spark.sql.files.maxPartitionBytes=1g\
 --conf spark.sql.shuffle.partitions=200\
 --conf spark.sql.adaptive.enabled=true\
 --conf spark.shuffle.manager=com.nvidia.spark.rapids.spark311.RapidsShuffleManager\
 --conf spark.shuffle.service.enabled=false\
 --conf spark.sql.broadcastTimeout=600\
 --conf spark.plugins=com.nvidia.spark.SQLPlugin\
 --conf spark.rapids.sql.concurrentGpuTasks=1\
 --conf spark.rapids.memory.pinnedPool.size=8G\
 --conf spark.rapids.sql.batchSizeBytes=1g\
 --conf spark.rapids.memory.gpu.direct.storage.spill.enabled=false\
 --conf spark.rapids.shuffle.transport.enabled=true\
 --conf spark.executorEnv.UCX_TLS=cuda_copy,cuda_ipc,rc,tcp\
 --conf spark.executorEnv.UCX_ERROR_SIGNALS=\
 --conf spark.executorEnv.UCX_MAX_RNDV_RAILS=1\
 --conf spark.executorEnv.UCX_MEMTYPE_CACHE=n\
 --conf spark.rapids.shuffle.maxMetadataSize=512K\
 --conf spark.executorEnv.UCX_RNDV_SCHEME=put_zcopy\
 --conf spark.executorEnv.UCX_RC_RX_QUEUE_LEN=1024\
 --conf spark.executorEnv.UCX_UD_RX_QUEUE_LEN=1024\
 --conf spark.driver.memory=10G\
 --conf spark.driver.maxResultSize=0\
 --conf spark.driver.extraJavaOptions=-Dai.rapids.cudf.nvtx.enabled=false\
 --conf spark.executor.extraClassPath=/opt/rapids/cudf.jar:/opt/rapids/rapids-4-spark.jar\
 --conf spark.executor.extraJavaOptions=-Dai.rapids.cudf.nvtx.enabled=false\
 --conf spark.executor.instances=1\
 --conf spark.executor.cores=24\
 --conf spark.executor.memory=64G\
 --conf spark.executor.resource.gpu.amount=1\
 --conf spark.task.cpus=1\
 --conf spark.task.resource.gpu.amount=0.0416\
 --jars /opt/rapids/cudf.jar,/opt/rapids/rapids-4-spark.jar,/opt/rapids/rapids-4-spark-benchmarks.jar
```
and the query:
```scala
import com.nvidia.spark.rapids.tests._
import com.nvidia.spark.rapids.tests.tpcds._

TpcdsLikeSpark.setupAllParquet(spark, "/opt/data/tpcds-1TB", true)

val benchmark = new BenchmarkRunner(new TpcdsLikeBench(true))
benchmark.collect(spark, "q1", 1)
```

**Expected behavior**
Should not throw exception.

**Environment details (please complete the following information)**
 - Environment location: Standalone

**Additional context**
Stacktrace:
```
java.lang.IllegalStateException: Close called too many times DeviceMemoryBuffer{address=0x7fc58663c900, length=6576072, id=5336}
        at ai.rapids.cudf.MemoryBuffer.close(MemoryBuffer.java:208)
        at com.nvidia.spark.rapids.LazySpillableGatherMapImpl.$anonfun$close$3(JoinGatherer.scala:360)
        at com.nvidia.spark.rapids.LazySpillableGatherMapImpl.$anonfun$close$3$adapted(JoinGatherer.scala:360)
        at scala.Option.foreach(Option.scala:407)
        at com.nvidia.spark.rapids.LazySpillableGatherMapImpl.close(JoinGatherer.scala:360)
        at com.nvidia.spark.rapids.JoinGathererImpl.close(JoinGatherer.scala:574)
        at org.apache.spark.sql.rapids.execution.HashJoinIterator.$anonfun$close$2(GpuHashJoin.scala:216)
        at org.apache.spark.sql.rapids.execution.HashJoinIterator.$anonfun$close$2$adapted(GpuHashJoin.scala:216)
        at scala.Option.foreach(Option.scala:407)
        at org.apache.spark.sql.rapids.execution.HashJoinIterator.close(GpuHashJoin.scala:216)
        at org.apache.spark.sql.rapids.execution.HashJoinIterator.$anonfun$new$1(GpuHashJoin.scala:222)
        at org.apache.spark.sql.rapids.execution.HashJoinIterator.$anonfun$new$1$adapted(GpuHashJoin.scala:222)
        at org.apache.spark.TaskContext$$anon$1.onTaskCompletion(TaskContext.scala:125)
        at org.apache.spark.TaskContextImpl.$anonfun$markTaskCompleted$1(TaskContextImpl.scala:124)
        at org.apache.spark.TaskContextImpl.$anonfun$markTaskCompleted$1$adapted(TaskContextImpl.scala:124)
        at org.apache.spark.TaskContextImpl.$anonfun$invokeListeners$1(TaskContextImpl.scala:137)
        at org.apache.spark.TaskContextImpl.$anonfun$invokeListeners$1$adapted(TaskContextImpl.scala:135)
        at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
        at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
        at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
        at org.apache.spark.TaskContextImpl.invokeListeners(TaskContextImpl.scala:135)
        at org.apache.spark.TaskContextImpl.markTaskCompleted(TaskContextImpl.scala:124)
        at org.apache.spark.scheduler.Task.run(Task.scala:141)
        at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:497)
        at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1439)
        at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:500)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```

### PR DESCRIPTION
**Pull request by rongou**
Closes #2348 

Signed-off-by: Rong Ou <rong.ou@gmail.com>

### CODE REVIEW

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsDeviceMemoryStore.scala [hunk 1]
```diff
@@ -67,22 +67,18 @@ class RapidsDeviceMemoryStore(catalog: RapidsBufferCatalog = RapidsBufferCatalog
...
+        Some(table),
+        contigBuffer,
+        initialSpillPriority,
+        spillCallback)) { buffer =>
       logDebug(s"Adding table for: [id=$id, size=${buffer.size}, " +
           s"meta_id=${buffer.meta.bufferMeta.id}, meta_size=${buffer.meta.bufferMeta.size}]")
       addBuffer(buffer)
-    } catch {
-      case t: Throwable =>
-        buffer.free()
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsDeviceMemoryStore.scala:84]: It needs to call `free` on the buffer, not `close`.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsDeviceMemoryStore.scala [hunk 2]
```diff
@@ -106,23 +102,19 @@ class RapidsDeviceMemoryStore(catalog: RapidsBufferCatalog = RapidsBufferCatalog
...
+        contigBuffer,
+        initialSpillPriority,
+        spillCallback)) { buffer =>
       logDebug(s"Adding table for: [id=$id, size=${buffer.size}, " +
           s"uncompressed=${buffer.meta.bufferMeta.uncompressedSize}, " +
           s"meta_id=${buffer.meta.bufferMeta.id}, meta_size=${buffer.meta.bufferMeta.size}]")
       addBuffer(buffer)
-    } catch {
-      case t: Throwable =>
-        buffer.free()
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsDeviceMemoryStore.scala:124]: Same `free` not `close` here.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsDeviceMemoryStore.scala [hunk 3]
```diff
@@ -141,24 +133,20 @@ class RapidsDeviceMemoryStore(catalog: RapidsBufferCatalog = RapidsBufferCatalog
...
+        initialSpillPriority,
+        spillCallback)) { buff =>
       logDebug(s"Adding receive side table for: [id=$id, size=${buffer.getLength}, " +
           s"uncompressed=${buff.meta.bufferMeta.uncompressedSize}, " +
           s"meta_id=${tableMeta.bufferMeta.id}, " +
           s"meta_size=${tableMeta.bufferMeta.size}]")
       addBuffer(buff)
-    } catch {
-      case t: Throwable =>
-        buff.free()
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsDeviceMemoryStore.scala:160]: Same `free` not `close` here.

**Review by jlowe — CHANGES_REQUESTED** [review]

**Comment by rongou** [comment]
Hmm why do we have separate `close` and `free` calls? When the refcount goes to 0, shouldn't a buffer be automatically freed?

**Comment by jlowe** [comment]
> why do we have separate close and free calls?

Because there's a semantic difference between "I'm done with this buffer" and "I want this buffer to release its contents ASAP", and I didn't want to rely on arbitrary users of the spill buffer to get the ref counting correct (it is quite easy to accidentally double-close).  If we go solely off the ref count then a user who accidentally double-closes ends up freeing the buffer too soon rather than getting a double-close error as they do today.

**Comment by rongou** [comment]
This seems pretty confusing as we are mixing metaphors with c-style `free()` and java-style `AutoCloseable`/try-with-resources. And the underlying `DeviceMemoryBuffer` does free on last close.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsDeviceMemoryStore.scala (hunk 1 above)
Comment (rongou) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsDeviceMemoryStore.scala:84]: Done.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsDeviceMemoryStore.scala (hunk 2 above)
Comment (rongou) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsDeviceMemoryStore.scala:124]: Done.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsDeviceMemoryStore.scala (hunk 3 above)
Comment (rongou) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsDeviceMemoryStore.scala:160]: Done.

**Review by rongou — COMMENTED** [review]
I'm not totally convinced. :) At the minimum we should not hide the underlying exception, making it harder to debug.

**Review by jlowe — APPROVED** [review]

**Comment by abellina** [comment]
build

**Comment by jlowe** [comment]
build

**Review by abellina — APPROVED** [review]

**Pull request merged**